criu/kerndat.c

```diff
@@ -822,19 +823,51 @@ static int kerndat_x86_has_ptrace_fpu_xsave_bug(void)
 	return 0;
 }
 
-#define KERNDAT_CACHE_FILE	KDAT_RUNDIR"/criu.kdat"
-#define KERNDAT_CACHE_FILE_TMP	KDAT_RUNDIR"/.criu.kdat"
+#define KERNDAT_CACHE_NAME	"criu.kdat"
+#define KERNDAT_CACHE_FILE	KDAT_RUNDIR"/"KERNDAT_CACHE_NAME
+
+static int get_kerndat_filename(char **kdat_file)
+{
+	int ret;
+
+	/*
+	 * Running as non-root, even with CAP_CHECKPOINT_RESTORE, does not
+	 * allow to write to KDAT_RUNDIR which usually is only writable by root.
+	 * Let's write .criu.kdat file to the home directory for non-root cases.
+	 */
+	if (opts.uid) {
```

Comment:
opts.unprivileged. opts.uid doesn't guarantee that criu has been executed with a full set of capabilities.